### ISSUE TITLE
Help with Jax, specifically complex differentiation!

### ISSUE THREAD

**Issue by hamza11235** [google/jax#4996]
Hi everyone,

I'm new here so my apologies if my query doesn't come across clearly. I'm currently doing computational project where I'm using JAX and specifically its complex autograd capabilities. The goal is to make the NN learn a unique function called the J-Invariant. So to update the parameters I am currently using grad, but what happens is that most of the time the gradients go to 0 after a single update. I think the problem is that I'm passing holomorphic=True even when I don;t know what's being passed is holomorphic or not. So I've decided to use the JVP function and simply differentiate along the 1+0*j vector. Now the problem I'm having currently is that I used stax to build up my neural network, and so my parameters have a specific pytree structure. JVP requires the primals (parameters) and the tangent vectors to be of the same structure. How do I create such an object such that each parameter variable is differentiated along the 1+0*j vector?
My apologies if this is a basic question; I'm a physics student and only recently got into machine learning, and no one I know seems to know much about JAX. I have attached my Python 3 file. Thank you in advance!

[The J-Invariant.zip](https://github.com/google/jax/files/5581041/The.J-Invariant.zip)

**Comment by mattjj** [comment]
Thanks for the question!

I think [`jax.tree_util.tree_map`](https://jax.readthedocs.io/en/latest/jax.tree_util.html?highlight=tree_map#jax.tree_util.tree_map) may be useful here. See also the [JAX autodiff cookbook section on complex numbers](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation).

Getting a directional derivative in the 1+0j direction won't necessarily help you do optimization though. If your function isn't holomorphic, maybe you just want to model it as having real-valued inputs, even if it uses complex numbers internally? (I'm assuming the loss value you're trying to optimize is real-valued.)

Could you share your example code as a [gist](https://gist.github.com), or a link to a colab notebook? Otherwise it'd be easier to look at a .py file rather than an .ipynb file.

**Comment by hamza11235** [comment]
Thank you for the response! Here's the link to the gist: [](https://gist.github.com/hamza11235/3e7fd033fa51c2dcb570dbacae965814)

I've thought about what you're suggesting about using two vectors, one with the real part and one with imaginary part. And that is definitely something I will try if this doesn't work. My loss function currently gives a complex output; i'm using a basic MSE loss at the moment . The reason I'm pushing on this is because my advisor insists that I use JAX and complex numbers since he wants to learn about NN with complex numbers in general, as it is important in other areas of his work. Also, assuming that I can model this function accurately using 2 real inputs, shouldn't this be equivalent to me using Vmap since this is unambiguous and holomorphy doesn't matter here? 

Let me give you a run down of what the goal is so you can maybe get a better idea. The J-invariant is a unique function that is defined in the upper half complex plane, and that is holomorphic (except at one point, z=i*infinity), and is invariant under Sl(2,z) transformations (or the modular group). Since this function is unique, if I can impose these conditions on a NN, I should get out outputs of the J-function (at least that's the idea). Also, an important point is that there's a "fundamental domain" of this function from which I can get all points in the complex plane from Sl(2,z) transformations of the points in the fundamental domain.

Coding-wise, I'm basically trying to impose two loss functions, one for the Sl(2,z) invariance and one for holomorphy. You can see these in the code. For Holomorphy, I'm basically using the CR relations. For my other loss, basically something like f(z)=f(z') is my loss function where z and z' are related by Sl(2,z). So I start with a random input from the fundamental domain, make a number of Sl(2,z) transformations on it (the 'trans' function in my code), use the parameters of the network to get an output for both my original complex number and my set of transformed complex numbers, and impose a loss function minimizing their difference. 

The problem I'm currently having is that in some runs, the gradient just drops to 0 after the first update. Do you have any idea why this could be happening? Also, my outputs at the moment don't match what I should be getting, but I want to fix this 0 gradients problem first and see if that fixes it. If you actually read through all that, thank you so much! Let me know if somethin is unclear in the code!

I've thought about what you're suggesting about using two vectors, one with the real part and one with imaginary part. And that is definitely something I will try if this doesn't work. My loss function currently gives a complex output; i'm using a basic MSE loss at the moment . The reason I'm pushing on this is because my advisor insists that I use JAX and complex numbers since he wants to learn about NN with complex numbers in general, as it is important in other areas of his work. Also, assuming that I can model this function accurately using 2 real inputs, shouldn't this be equivalent to me using Vmap since this is unambiguous and holomorphy doesn't matter here? #

Let me give you a run down of what the goal is so you can maybe get a better idea. The J-invariant is a unique function that is defined in the upper half complex plane, and that is holomorphic (except at one point, z=i*infinity), and is invariant under Sl(2,z) transformations (or the modular group). Since this function is unique, if I can impose these conditions on a NN, I should get out outputs of the J-function (at least that's the idea). Also, an important point is that there's a "fundamental domain" of this function from which I can get all points in the complex plane from Sl(2,z) transformations of the points in the fundamental domain.

Coding-wise, I'm basically trying to impose two loss functions, one for the Sl(2,z) invariance and one for holomorphy. You can see these in the code. For Holomorphy, I'm basically using the CR relations. For my other loss, basically something like f(z)=f(z') is my loss function where z and z' are related by Sl(2,z). So I start with a random input from the fundamental domain, make a number of Sl(2,z) transformations on it (the 'trans' function in my code), use the parameters of the network to get an output for both my original complex number and my set of transformed complex numbers, and impose a loss function minimizing their difference. 

The problem I'm currently having is that in some runs, the gradient just drops to 0 after the first update. Do you have any idea why this could be happening? Also, my outputs at the moment don't match what I should be getting, but I want to fix this 0 gradients problem first and see if that fixes it. If you actually read through all that, thank you so much! Let me know if something is unclear in the code!

**Comment by hamza11235** [comment]
Sorry, link didn't go through: https://gist.github.com/hamza11235/3e7fd033fa51c2dcb570dbacae965814

**Comment by hamza11235** [comment]
Thank you for the response! Here's the link to the gist: https://gist.github.com/hamza11235/3e7fd033fa51c2dcb570dbacae965814

I've thought about what you're suggesting about using two vectors, one with the real part and one with imaginary part. And that is definitely something I will try if this doesn't work. My loss function currently gives a complex output; i'm using a basic MSE loss at the moment . The reason I'm pushing on this is because my advisor insists that I use JAX and complex numbers since he wants to learn about NN with complex numbers in general, as it is important in other areas of his work. Also, assuming that I can model this function accurately using 2 real inputs, shouldn't this be equivalent to me using Vmap since this is unambiguous and holomorphy doesn't matter here?

Let me give you a run down of what the goal is so you can maybe get a better idea. The J-invariant is a unique function that is defined in the upper half complex plane, and that is holomorphic (except at one point, z=i*infinity), and is invariant under Sl(2,z) transformations (or the modular group). Since this function is unique, if I can impose these conditions on a NN, I should get out outputs of the J-function (at least that's the idea). Also, an important point is that there's a "fundamental domain" of this function from which I can get all points in the complex plane from Sl(2,z) transformations of the points in the fundamental domain.

Coding-wise, I'm basically trying to impose two loss functions, one for the Sl(2,z) invariance and one for holomorphy. You can see these in the code. For Holomorphy, I'm basically using the CR relations. For my other loss, basically something like f(z)=f(z') is my loss function where z and z' are related by Sl(2,z). So I start with a random input from the fundamental domain, make a number of Sl(2,z) transformations on it (the 'trans' function in my code), use the parameters of the network to get an output for both my original complex number and my set of transformed complex numbers, and impose a loss function minimizing their difference.

The problem I'm currently having is that in some runs, the gradient just drops to 0 after the first update. Do you have any idea why this could be happening? Also, my outputs at the moment don't match what I should be getting, but I want to fix this 0 gradients problem first and see if that fixes it. If you actually read through all that, thank you so much! Let me know if something is unclear in the code!

**Comment by mattjj** [comment]
Hmm, I'm not really familiar with the maths here, but here are some comments nonetheless!

> My loss function currently gives a complex output; i'm using a basic MSE loss at the moment

Shouldn't an MSE be real-valued? That is, the 'error' is usually the value of some norm (say the standard 2-norm on C^n) and norms are real-valued. You could write it like this:

```python
def loss(params, batch):
    inputs, targs = batch
    predictions = net_apply(params, inputs)
    targets=net_apply(params, targs)
    diff = predictions - targets
    return jnp.sum(jnp.vdot(diff, diff))
```

Notice the last two lines are different from the gist, and that `jnp.vdot` conjugates its first input so as to implement the standard inner product on complex vector spaces. 

In general, it makes sense to minimize a real-valued loss, but I'm not sure what it would mean to minimize a complex-valued one (since complex numbers aren't even ordered).

> Also, assuming that I can model this function accurately using 2 real inputs, shouldn't this be equivalent to me using Vmap since this is unambiguous and holomorphy doesn't matter here?

Yeah, though whether you use `vmap` just depends on whether you're representing the real- and imaginary-parts as stacked together in a single array or as separate arrays.

> The problem I'm currently having is that in some runs, the gradient just drops to 0 after the first update. Do you have any idea why this could be happening? 

I don't have any ideas at this point, but sorting out the types of things (like whether the loss should be real-valued) seems like a good start! Then finding minimal code that reproduces the issue might help isolate it.

**Comment by hamza11235** [comment]
Thank you for such a detailed response!

Yes, I agree, the loss should be real-valued. I think what I was trying to minimize was the difference between the 2 complex numbers themselves, but this is obviously the same as minimizing the magnitude of the difference vector, which is just the norm.

Also, I just realized I made a typo in the previous comment; I meant jvp, not vmap.

Also, thanks for the tree_map tip. That worked. However, I just realized after applying jvp (and getting an error) when the gradient calculated by jvp is passed to the optimizer. The error is due to the pytree structure of this gradient not being the same as the pytree structure of the parameters; the gradient is just a number. This is not what grad gives us. Why do you think jvp gives a number but grad gives us a pytree structure which is the same as the parameters (as it should)? And how can I remedy this? Thanks again!

**Comment by mattjj** [comment]
Well, a Jacobian-vector product gives a result that is the same shape/dimension as the output. A gradient has the same shape/dimension as the input, and is basically defined as `grad(f)(x) == vjp(f, x)[1](1.)`, i.e. as a vector-Jacobian product where the vector is a scalar 1.

**Comment by mattjj** [comment]
I think with a real-valued loss you should be able to use `grad` directly. To minimize a real-valued loss as a function of complex-valued inputs, you'd want to take steps in the negative elementwise-conjugate direction of the gradient (the reason is explained in the docs I linked above).

**Comment by hamza11235** [comment]
Hi, thank you for the clarification. 
So my loss function is now minimizing well, and I don't have pass holomorphic=True so no ambiguity there. I did not understand your comment about the taking steps in the conjugate direction though. How do I do that with grad? It doesn't specifically say in the documentation what that means. Also, do I even need to considering my loss is minimizing? Thanks again
